core/src/main/scala/kafka/server/BrokerServer.scala

```diff
@@ -93,6 +93,9 @@ class BrokerServer(
 
   this.logIdent = logContext.logPrefix
 
+  // VisibleForTesting
+  private[kafka] var controlledShutdownTimeoutMillis: Long = 5 * 60 * 1000L // 5 minutes
```

Comment:
This can solve the problem but is not very elegant, we can add a `shutdown(timeout)`, and use `Time` to compute a deadline, then we can use the deadline to compute a timeout for a controlled shutdown, WDYT?

Reply:
Thanks for reviewing this @dengziming. I agree it is not an elegant approach.
And thanks for your suggestion, it seems to be in line with @hachikuji 's suggestion as well.
I've updated this patch to follow your suggested approach. Please let me know if this is what you had in mind.
However, I'm not very confident that this is better, since `KafkaServer` also has to implement the new `shutdown` variant and it has no use for the timeout. So it seems a bit strange to have an implementation where the timeout has to be ignored.